rs/cli/src/idlgen.rs

```diff
@@ -0,0 +1,408 @@
+use anyhow::Context;
+use cargo_metadata::{camino::*, Package, PackageId};
+use std::{
+    collections::{HashMap, HashSet},
+    env, fs,
+    path::{Path, PathBuf},
+    process::{Command, ExitStatus},
+};
+
+pub struct CrateIdlGenerator {
+    manifest_path: Utf8PathBuf,
+    target_dir: Option<Utf8PathBuf>,
+    deps_level: Option<usize>,
+}
+
+impl CrateIdlGenerator {
+    pub fn new(
+        manifest_path: Option<PathBuf>,
+        target_dir: Option<PathBuf>,
+        deps_level: Option<usize>,
+    ) -> Self {
+        Self {
+            manifest_path: Utf8PathBuf::from_path_buf(
+                manifest_path.unwrap_or_else(|| env::current_dir().unwrap().join("Cargo.toml")),
+            )
+            .unwrap(),
+            target_dir: target_dir
+                .and_then(|p| p.canonicalize().ok())
+                .map(Utf8PathBuf::from_path_buf)
+                .and_then(|t| t.ok()),
+            deps_level,
+        }
+    }
+
+    pub fn generate(self) -> anyhow::Result<()> {
+        println!("...reading metadata: {}", &self.manifest_path);
+        // get metadata with deps
+        let metadata = cargo_metadata::MetadataCommand::new()
+            .manifest_path(&self.manifest_path)
+            .exec()?;
+
+        // find `sails-rs` packages (any version )
+        let sails_packages = metadata
+            .packages
+            .iter()
+            .filter(|&p| p.name == "sails-rs")
+            .collect::<Vec<_>>();
+
+        let target_dir = self
+            .target_dir
+            .as_ref()
+            .unwrap_or(&metadata.target_directory);
+
+        let package_list = get_package_list(&metadata, self.deps_level.unwrap_or(1))?;
```

Comment:
Do you want to make `deps_level` as usize in the struct and just initialize via `unwrap_or`